### PR TITLE
Add screenshot to search results

### PR DESCRIPTION
**Pull request by szakarias**

### CODE REVIEW

**Comment by szakarias** [comment]
image

File: app/lib/frontend/templates/views/pkg/package_list.dart [hunk 1]
```diff
@@ -45,8 +45,10 @@ d.Node _sdkLibraryItem(SdkLibraryHit hit) {
   ].join(' • ');
 
   return _item(
+    screenshots: [],
     url: hit.url!,
     name: hit.library!,
+    version: hit.version,
```
Comment (sigurdm) [inline app/lib/frontend/templates/views/pkg/package_list.dart:51]: This is not related to screenshots - right?

@isoos is there a reason we did not have the version check here before?

File: app/lib/package/models.dart [hunk 2]
```diff
@@ -855,25 +859,26 @@ class PackageView extends Object with FlagMixin {
...
+        flags: flags,
+        publisherId: publisherId,
+        isPending: isPending,
+        likes: likes,
+        grantedPubPoints: grantedPubPoints,
+        maxPubPoints: maxPubPoints,
+        tags: tags,
+        spdxIdentifiers: spdxIdentifiers,
+        apiPages: apiPages ?? this.apiPages,
+        screenshots: screenshots);
```
Comment (sigurdm) [inline app/lib/package/models.dart:881]: Nit: Add a trailing comma

**Review by sigurdm — APPROVED** [review]
LGTM

File: app/lib/frontend/templates/views/pkg/package_list.dart (hunk 1 above)
Comment (isoos) [reply app/lib/frontend/templates/views/pkg/package_list.dart:51]: It seems to be that this is required for the screenshot URL. Unfortunately, I don't think that `version` will be provided for Flutter SDK, but in such cases it won't matter as we will not have screenshots for such hits...

File: app/lib/frontend/templates/views/pkg/package_list.dart [hunk 3]
```diff
@@ -146,15 +150,25 @@ d.Node _packageItem(
...
   required String name,
+  required String? version,
   required DateTime? newTimestamp,
   required d.Node? labeledScoresNode,
   required String description,
   required d.Node metadataNode,
   required d.Node? tagsNode,
   required List<_ApiPageUrl>? apiPages,
 }) {
+  final bool hasScreenshots = screenshots != null && screenshots.isNotEmpty;
```
Comment (isoos) [inline app/lib/frontend/templates/views/pkg/package_list.dart:164]: maybe add `&& version != null` just in case, as the `getImageURL` uses `version!`

Comment (szakarias) [reply app/lib/frontend/templates/views/pkg/package_list.dart]: done

**Review by sigurdm — APPROVED** [review]
LGTM

File: app/lib/frontend/templates/views/pkg/package_list.dart [hunk 4]
```diff
@@ -189,12 +200,38 @@ d.Node _item({
...
+            d.img(
+                image: d.Image(
+                    alt: 'screenshot',
+                    width: null,
+                    height: null,
+                    src: thumbnailUrl)),
+            d.img(
+                classes: ['collections-icon'],
+                image: d.Image(
+                    height: null, // We handle sizing in css
```
Comment (isoos) [inline app/lib/frontend/templates/views/pkg/package_list.dart:229]: I think the lighthouse analysis will downscore us if we don't put width and height here. Let's just repeat them here too, event though the CSS may have them as well.

File: app/lib/package/screenshots/backend.dart [hunk 5]
```diff
@@ -67,4 +67,8 @@ class ImageStorage {
   }
 
   // TODO(zarah) add functionality to garbage collect bucket entries.
+
+  String getImageURL(String package, String version, String imageFilePath) {
+    return bucket.objectUrl([package, version, imageFilePath].join('/'));
```
Comment (isoos) [inline app/lib/package/screenshots/backend.dart:72]: Note to self: we may need to escape the `bucket.objectUrl` URLs.
